### PR TITLE
Fix timestamp issue

### LINKED ISSUE #704: Timestamp format needs to be updated in the conflict message displayed
In the message already contains result for timestamp - timestamp displayed is not in the same format as what was specified

Timestamp displayed - 2022-01-23 18:55:43.511, instead of 2022-01-23T18:55:43.511Z

### PR DESCRIPTION
**Pull request by khansaad**
This PR fixes #704 

- Design doc update to let the user know about the fixed timestamp format : **yyyy-MM-dd'T'HH:mm:ss'Z'**

### CODE REVIEW

File: src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java [hunk 1]
```diff
@@ -79,7 +87,7 @@ public void validate(List<ExperimentResultData> experimentResultDataList, Map<St
                         if (null != kruizeObject.getResultData())
                             isExist = kruizeObject.getResultData().contains(resultData);
                         if (isExist) {
-                            errorMsg = errorMsg.concat(String.format("Experiment name : %s already contains result for timestamp : %s", resultData.getExperiment_name(), resultData.getEndtimestamp()));
+                            errorMsg = errorMsg.concat(String.format("Experiment name : %s already contains result for timestamp : %s", resultData.getExperiment_name(), resultData.getEndtimestamp().toInstant()));
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java:90]: Why is this being converted to `toInstant`? Will the user see the same date String as what they sent in?

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java:90]: Yes. This is how the response is coming now:

```
{
    "message": "Experiment name : quarkus-resteasy-autotune-min-http-response-time-db already contains result for timestamp : 2022-01-26T06:40:48.511Z",
    "httpcode": 409,
    "documentationLink": "",
    "status": "ERROR"
}
```

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java:90]: It was missing the `T` and `Z` in the timestamp response while converting to string.

`"message": "Experiment name : quarkus-resteasy-autotune-min-http-response-time-db already contains result for timestamp : 2022-01-26 12:10:48.511",`

File: src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java [hunk 2]
```diff
@@ -79,7 +90,10 @@ public void validate(List<ExperimentResultData> experimentResultDataList, Map<St
                         if (null != kruizeObject.getResultData())
                             isExist = kruizeObject.getResultData().contains(resultData);
                         if (isExist) {
-                            errorMsg = errorMsg.concat(String.format("Experiment name : %s already contains result for timestamp : %s", resultData.getExperiment_name(), resultData.getEndtimestamp()));
+                            DateFormat dateFormat = new SimpleDateFormat(KruizeConstants.DateFormats.STANDARD_JSON_DATE_FORMAT);
+                            dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java:94]: Are we assuming that the input date is always in UTC timezone?

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java]: Updated now.

File: src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java (hunk 1 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java]: By using the `java.sql.Timestamp` class, we're able to get the Timezone correctly if there's no offset value i.e. if it's like this `2022-01-26T12:00:00.511Z` . But this fails for the cases when there's an offset like this added to it, `2022-01-26T12:00:00.511+03:30` . 

One of the solution I found and tested was if we switch to `java.time.OffsetDateTime` class from the  `Timestamp` class

File: src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java [hunk 3]
```diff
@@ -58,19 +58,19 @@ public void validate(List<ExperimentResultData> experimentResultDataList, Map<St
...
                             // Calculate the lower and upper bounds for the acceptable range i.e. +-5 seconds
-                            double lowerRange = (parsedMeasurementDuration * KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE - KruizeConstants.TimeConv.MEASUREMENT_DURATION_THRESHOLD_SECONDS) / KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE;
-                            double upperRange = (parsedMeasurementDuration * KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE + KruizeConstants.TimeConv.MEASUREMENT_DURATION_THRESHOLD_SECONDS) / KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE;
-                            if (!(durationInMins >= lowerRange && durationInMins <= upperRange)) {
+                            double lowerRange = Math.abs(parsedMeasurementDuration * KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE - KruizeConstants.TimeConv.MEASUREMENT_DURATION_THRESHOLD_SECONDS);
+                            double upperRange = (parsedMeasurementDuration * KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE + KruizeConstants.TimeConv.MEASUREMENT_DURATION_THRESHOLD_SECONDS);
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java:72]: Please use brackets to make the precedence clear and prevent any confusion

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/experiment/ExperimentResultValidation.java]: Updated now

**Comment by dinogun** [comment]
@khansaad Please fix the conflicts

**Comment by dinogun** [comment]
Closing stale PRs

**Pull request closed without merging**